io/src/main/scala/fs2/io/net/Socket.scala

```diff
@@ -69,3 +76,140 @@ trait Socket[F[_]] {
     */
   def writes: Pipe[F, Byte, INothing]
 }
+
+object Socket {
+  private[net] def forAsync[F[_]: Async](
+      ch: AsynchronousSocketChannel
+  ): Resource[F, Socket[F]] =
+    Resource.make {
+      (Semaphore[F](1), Semaphore[F](1)).mapN { (readSemaphore, writeSemaphore) =>
+        new AsyncSocket[F](ch, readSemaphore, writeSemaphore)
+      }
+    }(_ => Async[F].delay(if (ch.isOpen) ch.close else ()))
+
+  private final class AsyncSocket[F[_]](
+      ch: AsynchronousSocketChannel,
+      readSemaphore: Semaphore[F],
+      writeSemaphore: Semaphore[F]
+  )(implicit F: Async[F])
+      extends Socket[F] {
+    private[this] final val defaultReadSize = 8192
+    private[this] var readBuffer: ByteBuffer = ByteBuffer.allocateDirect(defaultReadSize)
+
+    private def getBufferForRead(size: Int): F[ByteBuffer] = F.delay {
+      if (readBuffer.capacity() < size)
+        readBuffer = ByteBuffer.allocateDirect(size)
+      else
+        (readBuffer: Buffer).limit(size)
+      readBuffer
+    }
```

Comment:
This would make clearer that all alterations to the `readBuffer` are under the `readSemaphore`. 
```suggestion
    private def withReadBuffer[A](size: Int)(k: ByteBuffer => F[A]): F[A] = 
      readSemaphore.permit.use { _ => F.delay {
        if (readBuffer.capacity() < size)
          readBuffer = ByteBuffer.allocateDirect(size)
        else
          (readBuffer: Buffer).limit(size)
        k(readBuffer)
      }.flatten
    }
```